README
                   ftrace - A function call tracer

$Id: README,v 1.6 2007/06/13 04:38:53 hamano Exp $

ftrace is a program that simply runs the specified command until it
exits. It intercepts and records the user define function calls which are
called by the executed process. 
This command should be compiled with -g and -finstrument-functions
flags.

* Requirements
  libglib-1.2
  libelf
  
* Build
  $ ./configure
  $ make
  # make install

* Usage
  ftrace [ command [ arg ...  ] ]


NEWS
See git log or README.md


uftrace-test.c
#include <stdio.h>
#include <pthread.h>
#include <string.h>

void foobar()
{
}

void foo()
{
}

void bar()
{   
    foobar();
    foobar();
}

void tf1()
{
}

void tf2()
{
}

void arg_int3(int x, int y, int z)
{
/*
    printf("&x: %p\n", &x);
    printf("&y: %p\n", &y);
    printf("&z: %p\n", &z);
*/
}
void arg_int2(int a, int b)
{
    arg_int3(a, b, b+1);
}
void arg_int1(int i)
{
    arg_int2(i, i+1);
}

void arg_uint3(unsigned int x, unsigned int y, unsigned int z)
{
}
void arg_uint2(unsigned int a, unsigned int b)
{
    arg_uint3(a, b, b+1);
}
void arg_uint1(unsigned int i)
{
    arg_uint2(i, i+1);
}

void arg_char3(char x, char y, char z)
{
}
void arg_char2(char a, char b)
{
    arg_char3(a, b, b+1);
}
void arg_char1(char i)
{
    arg_char2(i, i+1);
}

void arg_short3(short x, short y, short z)
{
}
void arg_short2(short a, short b)
{
    arg_short3(a, b, b+1);
}
void arg_short1(short i)
{
    arg_short2(i, i+1);
}

void arg_long3(long x, long y, long z)
{
}
void arg_long2(long a, long b)
{
    arg_long3(a, b, b+1);
}
void arg_long1(long i)
{
    arg_long2(i, i+1);
}
void arg_ulong3(unsigned long x, unsigned long y, unsigned long z)
{
}
void arg_ulong2(unsigned long a, unsigned long b)
{
    arg_ulong3(a, b, b+1);
}
void arg_ulong1(long i)
{
    arg_ulong2(i, i+1);
}

void arg_longlong3(long long x, long long y, long long z)
{
}
void arg_longlong2(long long a, long long b)
{
    arg_longlong3(a, b, b+1);
}
void arg_longlong1(long long i)
{
    arg_longlong2(i,i+1);
}
void arg_ulonglong3(unsigned long long x,
                    unsigned long long y,
                    unsigned long long z)
{
}
void arg_ulonglong2(unsigned long long a, unsigned long long b)
{
    arg_ulonglong3(a, b, b+1);
}
void arg_ulonglong1(unsigned long long i)
{
    arg_ulonglong2(i,i+1);
}

void arg_test(int x, char *z)
{
}

void *workthread(void* param)
{   
    tf1();
    tf2();
    pthread_exit(param);
}

int main(int argc, char *argv[])
{
    void *tresult;
    pthread_t thread=0;
    if(argc==2){
        if(!strcmp(argv[1],"-f")){
            if(!fork()){
                foobar();
            }
        }
        if(!strcmp(argv[1],"-t")){
            if(pthread_create(&thread, NULL, workthread, NULL)){
                printf("pthread_create error!\n");
            }else{
                pthread_join(thread, &tresult);
            }
        }
    }
    arg_int1(10);
    arg_uint1(10);
    arg_char1(20);
    arg_short1(30);
    arg_long1(40);
    arg_ulong1(50);
    arg_longlong1(60);
    arg_ulonglong1(70);
    arg_test(100, NULL);
    return 0;
}


ftrace.c
/*
 * ftrace.c
 * $Id: ftrace.c,v 1.10 2007/06/15 13:24:54 hamano Exp $
 * This program can be distributed under the terms of the GNU GPL.
 * See the file COPYING.
*/
#define _GNU_SOURCE
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <time.h>
#include <syslog.h>
#include <dlfcn.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>

void usage()
{
    printf("ftrace version %s\n", VERSION);
    printf("usage: ftrace [-s] [-o file] [-t] [-T] [-l] [-f regex_pattern] command [arg ...]\n");
    printf("option\n");
    printf("  -s   : output to syslog\n");
    printf("  -o   : output to file   (FileName: file.pid)\n");
    printf("  -t   : with '-o' option (FileName: file.pid.thread_id)\n");
    printf("  -T   : thread_id added to the output format of pid.\n");
    printf("  -l   : file name and line no. added to the output\n");
    printf("  -f   : select non-matching output \n");
    printf("\n");
}

int main(int argc,char *argv[])
{
    int i,j;
    char so[PATH_MAX];

    sprintf(so, "%s/libftrace.so", FTRACE_LIB);
    setenv("LD_PRELOAD", so, 1);

    if(argc == 1){
        usage();
        exit(0);
    }
    for(i=1;i<argc;i++){
        if(argv[i][0] == '-'){
            switch(argv[i][1]){
            case 's':
                unsetenv("FTRACE_PATH");
                setenv("FTRACE_OUTPUT", "syslog", 1);
                break;
            case 'o':
                setenv("FTRACE_PATH",   argv[++i], 1);
                setenv("FTRACE_OUTPUT", "file",    1);
                break;
            case 't':
                setenv("FTRACE_SPLIT_THREAD_S","1",1);
                break;
            case 'T':
                setenv("FTRACE_SPLIT_THREAD_L","1",1);
                break;
            case 'l':
                setenv("FTRACE_PRINT_FILENAME","1",1);
                break;
            case 'f':
                setenv("FTRACE_FILTER_IGNORE","1",1);
                setenv("FTRACE_FILTER_IGNORE_PATTERN",   argv[++i], 1);
                break;
            case 'h':
                usage();
                return(0);
            }
        }else{
            execvp(argv[i], argv + i);
            fprintf(stderr,"cant exec %s\n", argv[i]);
            exit(1);
        }
    }
    usage();
    return(1);
}


uftrace.c
/*
 * ftrace.c => rename uftrace.c
 * $Id: ftrace.c,v 1.10 2007/06/15 13:24:54 hamano Exp $
 * This program can be distributed under the terms of the GNU GPL.
 * See the file COPYING.
*/
#define _GNU_SOURCE
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <time.h>
#include <syslog.h>
#include <dlfcn.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>

void usage()
{
    printf("uftrace version %s\n", VERSION);
    printf("usage: uftrace [-s] [-o file] [-t] [-T] [-l] [-f regex_pattern] command [arg ...]\n");
    printf("option\n");
    printf("  -s   : output to syslog\n");
    printf("  -o   : output to file   (FileName: file.pid)\n");
    printf("  -t   : with '-o' option (FileName: file.pid.thread_id)\n");
    printf("  -T   : thread_id added to the output format of pid.\n");
    printf("  -l   : file name and line no. added to the output\n");
    printf("  -f   : select non-matching output \n");
    printf("  -e   : output info when exit function \n");
    printf("\n");
}

int main(int argc,char *argv[])
{
    int i,j;
    char so[PATH_MAX];

    sprintf(so, "%s/libuftrace.so", FTRACE_LIB);
    setenv("LD_PRELOAD", so, 1);

    if(argc == 1){
        usage();
        exit(0);
    }
    for(i=1;i<argc;i++){
        if(argv[i][0] == '-'){
            switch(argv[i][1]){
            case 's':
                unsetenv("FTRACE_PATH");
                setenv("FTRACE_OUTPUT", "syslog", 1);
                break;
            case 'o':
                setenv("FTRACE_PATH",   argv[++i], 1);
                setenv("FTRACE_OUTPUT", "file",    1);
                break;
            case 't':
                setenv("FTRACE_SPLIT_THREAD_S","1",1);
                break;
            case 'T':
                setenv("FTRACE_SPLIT_THREAD_L","1",1);
                break;
            case 'l':
                setenv("FTRACE_PRINT_FILENAME","1",1);
                break;
            case 'f':
                setenv("FTRACE_FILTER_IGNORE","1",1);
                setenv("FTRACE_FILTER_IGNORE_PATTERN",   argv[++i], 1);
                break;
            case 'e':
                setenv("FTRACE_EXIT_OUTPUT","1",1);
                break;
            case 'h':
                usage();
                return(0);
            }
        }else{
            execvp(argv[i], argv + i);
            fprintf(stderr,"cant exec %s\n", argv[i]);
            exit(1);
        }
    }
    usage();
    return(1);
}


libuftrace.c
/*
 * libftrace.c => rename to libuftrace.c
 * $Id: libftrace.c,v 1.6 2007/06/19 01:33:43 hamano Exp $
 * This program can be distributed under the terms of the GNU GPL.
 * See the file COPYING.
*/
#define _GNU_SOURCE
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <time.h>
#include <syslog.h>
#include <dlfcn.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <glib.h>
#include <glib/ghash.h>
#include <bfd.h>
#include <regex.h>
#include <errno.h>

#include <libuftrace.h>
#include "prototype.h"
#include <mcheck.h>


GHashTable *functions;
FILE *fp;
int opt_syslog = 0;
int opt_time = 1;
int opt_pid = 1;
int opt_threadid = 0;
int opt_type = 0;
int opt_filename = 0;
int opt_ignore = 0;
int opt_exit = 0;
int opt_start_depth = 0;
int opt_max_depth = 0;

static bfd *pbfd = NULL;
static asymbol **symbols = NULL;
static int nsymbol = 0;

regex_t preg;

int start_depth = 0;
int max_depth = 0;
int cur_depth = 0;


void ftrace_append_time(GString *line){
    struct timeval  tv;
    struct timezone tz;
    struct tm t;

    gettimeofday(&tv,&tz);
    localtime_r(&(tv.tv_sec), &t);
    g_string_append_printf(line, "%02d:%02d:%02d.%06ld ",
                      t.tm_hour, t.tm_min, t.tm_sec, tv.tv_usec);
}

void ftrace_append_pid(GString *line){
    struct timeval  tv;
    struct timezone tz;
    struct tm t;

    gettimeofday(&tv,&tz);
    localtime_r(&(tv.tv_sec), &t);

    if(opt_threadid){
        g_string_append_printf(line, "[%05d.0x%lx] ", getpid(),  (unsigned long)pthread_self());
    }else{
        g_string_append_printf(line, "[%05d] ", getpid());
    }
}

int ignore_filter_match(const char* pattern) {
    return !(REG_NOMATCH == regexec(&preg, pattern, 0, NULL, 0));
}

int ftrace_append_filename(GString *line, void* address){
    const char *file_name;
    const char *function_name;
    int lineno;
    
    asection *section = bfd_get_section_by_name(pbfd, ".debug_info");
    int found = bfd_find_nearest_line(pbfd, section, symbols, (long)address, 
                                  &file_name, &function_name, &lineno);
    if(found && file_name != NULL && function_name != NULL) {
        if(opt_ignore && ignore_filter_match(function_name)) return 0;
        g_string_append_printf(line, "(%s:%5d) ", basename(file_name), lineno);
    }
    return 1;
}

#if defined(__i386) || defined(__i386__) || \
defined(__powerpc) || defined(__powerpc__)

void *ftrace_append_arg(GString *line, arg_t *arg, void *frame){

    switch(arg->type){
    case TYPE_POINTER:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%p", frame);
        frame+=4;
        break;
    case TYPE_INT:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%d", *(int*)frame);
        frame+=4;
        break;
    case TYPE_UINT:
        if(opt_type) g_string_append_printf(line, "(%p)", arg->typename);
        g_string_append_printf(line, "%u", *(unsigned int*)frame);
        frame+=4;
        break;
    case TYPE_CHAR:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%d", *(char*)frame);
        frame+=4;
        break;
    case TYPE_UCHAR:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%u", *(unsigned char*)frame);
        frame+=4;
        break;
    case TYPE_SHORT:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%d", *(short*)frame);
        frame+=4;
        break;
    case TYPE_USHORT:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%u", *(unsigned short*)frame);
        frame+=4;
        break;
    case TYPE_LONG:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%ld", *(long*)frame);
        frame+=4;
        break;
    case TYPE_ULONG:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%ld", *(unsigned long*)frame);
        frame+=4;
        break;
    case TYPE_LONGLONG:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%lld",  *(long long*)frame);
        frame+=8;
        break;
    case TYPE_ULONGLONG:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%lld",  *(unsigned long long*)frame);
        frame+=8;
        break;
    default:
        g_string_append_printf(line, "unknown, ...");
        return NULL;
    }
    return frame;
}

#elif defined(__x86_64) || defined(__x86_64__)

void *ftrace_append_arg(GString *line, arg_t *arg, void *frame){

    switch(arg->type){
    case TYPE_POINTER:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%p", frame);
        frame-=4;
        break;
    case TYPE_INT:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%d", *(int*)frame);
        frame-=4;
        break;
    case TYPE_UINT:
        if(opt_type) g_string_append_printf(line, "(%p)", arg->typename);
        g_string_append_printf(line, "%u", *(unsigned int*)frame);
        frame-=4;
        break;
    case TYPE_CHAR:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%d", *(char*)frame);
        frame-=4;
        break;
    case TYPE_UCHAR:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%u", *(unsigned char*)frame);
        frame-=4;
        break;
    case TYPE_SHORT:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%d", *(short*)frame);
        frame-=4;
        break;
    case TYPE_USHORT:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%u", *(unsigned short*)frame);
        frame-=4;
        break;
    case TYPE_LONG:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%ld", *(long*)frame);
        frame-=8;
        break;
    case TYPE_ULONG:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%ld", *(unsigned long*)frame);
        frame-=8;
        break;
    case TYPE_LONGLONG:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%lld",  *(long long*)frame);
        frame-=8;
        break;
    case TYPE_ULONGLONG:
        if(opt_type) g_string_append_printf(line, "(%s)", arg->typename);
        g_string_append_printf(line, "%lld",  *(unsigned long long*)frame);
        frame-=8;
        break;
    default:
        g_string_append_printf(line, "unknown, ...");
        return NULL;
    }

    return frame;
}

#else

void *ftrace_append_arg(GString *line, arg_t *arg, void *frame){
    g_string_append_printf(line, "unknown, ...");
    return NULL;
}

#endif

int ftrace_append_function(GString *line, void *addr){
    function_t *func;
    char *name;
    void *frame;
    GSList *args;
    arg_t *arg;


#if defined(__i386) || defined(__i386__) || \
defined(__powerpc) || defined(__powerpc__)
    frame = __builtin_frame_address(3) + 8;
#elif defined(__x86_64) || defined(__x86_64__)
    frame = __builtin_frame_address(3) - 4;
#else
    frame = __builtin_frame_address(3) + 8;
#endif

    func = (function_t*)g_hash_table_lookup(functions, &addr);
    if(!func){
        g_string_append_printf(line, "UNKNOWN<0x%lx>()", (unsigned long)addr);
        return;
    }

    if(opt_ignore && ignore_filter_match(func->name)) return 0;

    g_string_append_printf(line, "%s(", func->name);
    args = func->args;
    while(args){
        arg = args->data;
        //fprintf(stderr, "%s %s %d\n", arg->typename, arg->name, arg->size);
#if defined(__x86_64) || defined(__x86_64__)
        if(args == func->args && \
           ( \
             arg->type == TYPE_LONG || \
             arg->type == TYPE_ULONG || \
             arg->type == TYPE_LONGLONG || \
             arg->type == TYPE_ULONGLONG )\
            ){
            frame -= 4;
        }
#endif
        frame = ftrace_append_arg(line, arg, frame);
        if(!frame){
            break;
        }
        args = g_slist_next(args);
        if(args){
            g_string_append_printf(line, ", ");
        }
    }
    g_string_append_printf(line, ")");
    return 1;
}

void init_symbols(){
    int size;
    pbfd = bfd_openr("/proc/self/exe", NULL);
    bfd_check_format(pbfd, bfd_object);
    size = bfd_get_symtab_upper_bound(pbfd);
    symbols = (asymbol **)malloc(size);
    nsymbol = bfd_canonicalize_symtab(pbfd, symbols);
}

void init_ignore(const char* pattern){
    int err = regcomp(&preg, pattern, REG_EXTENDED|REG_ICASE|REG_NOSUB|REG_NEWLINE);
    if(err) {
        perror("regex compile");
        exit(-1);
    }
}

void init_start_depth(const char* depth) {
    start_depth = (int)strtol(depth, NULL, 0);
    if(start_depth == EINVAL || start_depth == ERANGE) {
        perror("start_depth");
        exit(-1);
    }
}

void init_max_depth(const char* depth) {
    max_depth = (int)strtol(depth, NULL, 0);
    if(max_depth == EINVAL || max_depth == ERANGE) {
        perror("max_depth");
        exit(-1);
    }
}

void __attribute__((constructor))ftrace_init()
{
    const char *dfpath = "./uftrace_output";
    const char *target = getenv("FTRACE_OUTPUT");
    const char *ftpath = getenv("FTRACE_PATH");
    const char *stflag = getenv("FTRACE_SPLIT_THREAD_S");
    const char *dtflag = getenv("FTRACE_SPLIT_THREAD_L");
    const char *dfnameflag = getenv("FTRACE_PRINT_FILENAME");
    const char *ignoreflag = getenv("FTRACE_FILTER_IGNORE");
    const char *ignorepat  = getenv("FTRACE_FILTER_IGNORE_PATTERN");
    const char *exitflag  = getenv("FTRACE_EXIT_OUTPUT");
    const char *start_depth_flag  = getenv("FTRACE_FILTER_START_DEPTH");
    const char *start_depth_env  = getenv("FTRACE_START_DEPTH");
    const char *max_depth_flag  = getenv("FTRACE_FILTER_MAX_DEPTH");
    const char *max_depth_env  = getenv("FTRACE_MAX_DEPTH");
    char fname[PATH_MAX];

    if(target){
        if(!strcmp(target,"syslog")){
            opt_syslog=1;
            openlog("ftrace", LOG_CONS | LOG_PERROR, LOG_USER);
        }else{
            opt_syslog=0;
            if(!strcmp(target, "file")) {
                if(!ftpath){
                    ftpath = dfpath;
                }
                if(stflag){
                    sprintf(fname,"%s.%d.%u", ftpath, getpid(), pthread_self());
                }else{
                    sprintf(fname,"%s.%d", ftpath, getpid());
                }
                fp = fopen(fname,"a");
                if(fp == NULL){
                    fp = stderr;
                }
            } else {
                fp = stderr;
            }
        }
    }else{
        opt_syslog=0;
        fp = stderr;
    }

    if(dtflag) {
        opt_threadid = 1;
    }

    if(dfnameflag) {
        opt_filename = 1;
        init_symbols();
    }

    if(ignoreflag) {
        opt_ignore = 1;
        init_ignore(ignorepat);
    }

    if(exitflag) {
        opt_exit = 1;
    }

    if(start_depth_flag) {
        opt_start_depth = 1;
        init_start_depth(start_depth_env);
    }

    if(max_depth_flag) {
        opt_max_depth = 1;
        init_max_depth(max_depth_env);
    }

    functions = g_hash_table_new((GHashFunc)g_int_hash,
                                 (GCompareFunc)g_int_equal);
    prototype_init(functions, "/proc/self/exe");
    //prototype_print(functions);
}

void __attribute__((destructor))ftrace_finish()
{
    
    prototype_finish(functions);
    g_hash_table_destroy(functions);
    fclose(fp);
}

void cyg_profile_func_generic(void *this, void *callsite, const char* ent_or_exit)
{
    GString *line = g_string_new(NULL);
    int print_ok = 1;

    if(opt_time){
        ftrace_append_time(line);
    }
    if(opt_pid){
        ftrace_append_pid(line);
    }
    if(opt_filename) {
        print_ok = ftrace_append_filename(line, this);
    }
    if(opt_exit) {
        g_string_append_printf(line, "%s", ent_or_exit);
    }

    print_ok = ftrace_append_function(line, this);

    if(print_ok && opt_syslog){
        syslog(LOG_INFO, "%s", line->str);
    }else if(print_ok && !(opt_syslog)){
        g_string_append_c(line, '\n');
        fprintf(fp, "%s", line->str);
        fflush(fp);
    }
}


void __cyg_profile_func_enter(void *this, void *callsite)
{
    cyg_profile_func_generic(this, callsite, "enter ");
}

void __cyg_profile_func_exit(void *this, void *callsite)
{
    if(!opt_exit) return;
    cyg_profile_func_generic(this, callsite, "exit  ");
}

pid_t fork()
{
    pid_t pid;
    pid_t (*fork_org)(void);
    GString *line = g_string_new(NULL);

    if(opt_time){
        ftrace_append_time(line);
    }
    if(opt_pid){
        ftrace_append_pid(line);
    }
    fork_org = dlsym(RTLD_NEXT,"fork");
    pid = (*fork_org)();
    if(opt_time){
        ftrace_append_time(line);
    }
    if(opt_pid){
        ftrace_append_pid(line);
    }    
    g_string_append_printf(line, "fork() pid=%d", pid);
    g_string_append_c(line, '\n');
    fprintf(fp, "%s", line->str);
    fflush(fp);
    return(pid);
}

int pthread_create(pthread_t *thread, pthread_attr_t * attr,
                   void * (*start_routine)(void *), void * arg)
{
    int result;
    GString *line = g_string_new(NULL);
    int (*pthread_create_org)(pthread_t * thread, pthread_attr_t * attr,
                              void * (*start_routine)(void *), void * arg);
    pthread_create_org = dlsym(RTLD_NEXT,"pthread_create");

    if(opt_time){
        ftrace_append_time(line);
    }
    if(opt_pid){
        ftrace_append_pid(line);
    }
    result=(*pthread_create_org)(thread, attr, start_routine, arg);
    if(thread){
        g_string_append_printf(line, "%-16s: result=%d TID=%lu",
                          "pthread_create()", result, (unsigned long)*thread);
    }else{
        g_string_append_printf(line, "%-16s: result=%d       ",
                          "pthread_create()", result);
    }
    g_string_append_c(line, '\n');
    fprintf(fp,"%s", line->str);
    fflush(fp);
    return(result);
}

void pthread_exit(void *retval)
{
    GString *line = g_string_new(NULL);
    void (*pthread_exit_org)(void *retval);
    pthread_exit_org = dlsym(RTLD_NEXT,"pthread_exit");

    if(opt_time){
        ftrace_append_time(line);
    }
    if(opt_pid){
        ftrace_append_pid(line);
    }
    g_string_append_printf(line, "%-16s: retval=%p", "pthread_exit()", retval);
    g_string_append_c(line, '\n');
    fprintf(fp, "%s", line->str);
    fflush(fp);
    (*pthread_exit_org)(retval);
    exit(0);
}
